File: Solutions/2479-meeting-rooms-iii/meeting-rooms-iii.dart
class Solution {
  int mostBooked(int n, List<List<int>> meetings) {
    // List to keep track of when each room becomes available
    List<int> roomAvailabilityTime = List.filled(n, 0);

    // List to count how many meetings each room has handled
    List<int> meetingCount = List.filled(n, 0);

    // Sort meetings by their start time
    meetings.sort((a, b) => a[0].compareTo(b[0]));

    for (var meeting in meetings) {
      int start = meeting[0];
      int end = meeting[1];

      // Track the soonest available room (in case all are busy)
      int minRoomAvailabilityTime = 1 << 60; // A large number (simulating infinity)
      int minAvailableTimeRoom = 0;

      // Flag to indicate if a room was found that can host the meeting immediately
      bool foundUnusedRoom = false;

      // Try to find a room that is free by the start of the meeting
      for (int i = 0; i < n; i++) {
        if (roomAvailabilityTime[i] <= start) {
          // Room is free for use
          foundUnusedRoom = true;
          roomAvailabilityTime[i] = end; // Book the room until meeting end
          meetingCount[i]++; // Increase its meeting count
          break; // Exit the loop since we've found a free room
        }

        // Track the room that becomes free the soonest
        if (roomAvailabilityTime[i] < minRoomAvailabilityTime) {
          minRoomAvailabilityTime = roomAvailabilityTime[i];
          minAvailableTimeRoom = i;
        }
      }

      // If no room was immediately available, delay the meeting
      if (!foundUnusedRoom) {
        // Schedule the meeting to start after the room becomes free
        roomAvailabilityTime[minAvailableTimeRoom] += (end - start);
        meetingCount[minAvailableTimeRoom]++;
      }
    }

    // Find the room with the highest number of meetings
    int maxMeetings = meetingCount[0];
    int resultRoom = 0;

    for (int i = 1; i < n; i++) {
      if (meetingCount[i] > maxMeetings) {
        maxMeetings = meetingCount[i];
        resultRoom = i;
      }
    }

    // Return the room number that handled the most meetings
    return resultRoom;
  }
}

File: Solutions/2479-meeting-rooms-iii/meeting-rooms-iii.ts
function mostBooked(n: number, meetings: number[][]): number {
    // Sort meetings by their original start time
    meetings.sort((a, b) => a[0] - b[0]);

    // Count how many meetings each room holds
    const meetingCount: number[] = new Array(n).fill(0);

    // -------------------------------
    // Min-heap for available rooms
    // Stores room numbers
    // -------------------------------
    const availableRooms: number[] = [];
    for (let i = 0; i < n; i++) {
        availableRooms.push(i);
    }

    // Helper function to keep availableRooms as a min-heap
    availableRooms.sort((a, b) => a - b);

    // -------------------------------
    // Min-heap for busy rooms
    // Each element: [endTime, roomNumber]
    // Sorted by endTime first, then roomNumber
    // -------------------------------
    const busyRooms: [number, number][] = [];

    // Helper function to push into busyRooms heap
    const pushBusy = (item: [number, number]) => {
        busyRooms.push(item);
        busyRooms.sort((a, b) => {
            if (a[0] !== b[0]) return a[0] - b[0];
            return a[1] - b[1];
        });
    };

    // Helper function to pop from busyRooms heap
    const popBusy = () => busyRooms.shift()!;

    // Helper function to push into availableRooms heap
    const pushAvailable = (room: number) => {
        availableRooms.push(room);
        availableRooms.sort((a, b) => a - b);
    };

    // -----------------------------------
    // Process each meeting one by one
    // -----------------------------------
    for (const [start, end] of meetings) {
        const duration = end - start;

        // Free up rooms that have finished by 'start'
        while (busyRooms.length > 0 && busyRooms[0][0] <= start) {
            const [_, room] = popBusy();
            pushAvailable(room);
        }

        // If we have a free room
        if (availableRooms.length > 0) {
            // Use the lowest-numbered available room
            const room = availableRooms.shift()!;
            meetingCount[room]++;
            pushBusy([end, room]);
        } 
        // Otherwise, all rooms are busy → delay the meeting
        else {
            // Get the room that finishes earliest
            const [earliestEnd, room] = popBusy();

            // Meeting starts when this room becomes free
            const newEnd = earliestEnd + duration;

            meetingCount[room]++;
            pushBusy([newEnd, room]);
        }
    }

    // -----------------------------------
    // Find the room with the most meetings
    // -----------------------------------
    let maxMeetings = 0;
    let answerRoom = 0;

    for (let i = 0; i < n; i++) {
        if (meetingCount[i] > maxMeetings) {
            maxMeetings = meetingCount[i];
            answerRoom = i;
        }
    }

    return answerRoom;
}
